### PR TITLE
rgw:  support store head and tail objects separately

### PR DESCRIPTION
**Pull request by fangyuxiangGL**
We ever faced a problem when created radosgw object storage on ec pool(k=4, m=2).  Using cosbench to do upload test, we got 700-800 ops of 4K size objects,  but 2212.18 ops of 4K size objects from 3 replicated pool. Performance from ec is lower than 3 replicated pool, but they eventually have similar throughput when object size become bigger(4MB, 8MB or bigger). This phenomena is easy to explain, cpu is the bottleneck when upload small objects, but disks will be the bottleneck when upload bigger objects.

Our customers always concern cost, so ec is a good choice to lower the cost of capacity; but it also brings trouble as mentioned above. So I wanted to find a way to improve the performance of ec for object storage based radosgw, and found a way to leverage capacity and performance.

My opinion is that we should support store head and tail objects of radosgw object separately, which means that stores head objects in 3 replicated pool and tail objects in ec pool. So for small objects, we
can get performance of 3 replicated pool, and we also benefit 67% capacity utility from ec(3 replicated only has 33%).

Consider a scene: we want to upload big size(MB or GB) objects , we prefer to use multipart, radosgw will stripe every part to several tail rados objects but no head object and all of them will land in ec
pool. So we will get similar throughput as 3 replicated pool for they are big objects, and we also benefit capacity utility.

Pareto principle (also known as the 80–20 rule) generally exists in workload, which means 20% files/objects occupy 80% capacity. It is not just subjective guess,  my company's share disk(like dropbox, storing
department e-doc, software, and so on) obey the rule and even 85-15(15% files occupy 85% capacity).

### CODE REVIEW

File: src/rgw/rgw_common.h [hunk 1]
```diff
@@ -1917,6 +1934,7 @@ struct rgw_obj {
   rgw_obj_key key;
 
   bool in_extra_data{false}; /* in-memory only member, does not serialize */
+  bool head_obj{true}; /* in-memory only member, does not serialize */
```
Comment (yehudasa) [inline src/rgw/rgw_common.h]: @fangyuxiangGL `rgw_obj` represents as it's name an rgw object. This means that it represents a set of both head and tail. This member implies that you aim to use it to point at only the head part, which doesn't make sense.

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -121,17 +121,28 @@ static string RGW_DEFAULT_PERIOD_ROOT_POOL = "rgw.root";
...
                                   const string& placement_id, const rgw_obj& obj, rgw_pool *pool)
 {
-  if (!zone_params.get_head_data_pool(placement_id, obj, pool)) {
-    RGWZonePlacementInfo placement;
-    if (!zone_params.get_placement(zonegroup.default_placement, &placement)) {
-      return false;
-    }
+  bool find;
+
+  if (obj.is_head_obj())
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: @fangyuxiangGL seems to me that you'd want to audit the callers, and see whether they need to access the head or the tail and change behavior accordingly.

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -10053,12 +10067,19 @@ int RGWRados::Object::Read::read(int64_t ofs, int64_t end, bufferlist& bl)
...
+  } else {
+    store->open_pool_ctx(read_obj.pool, tail_io_ctx);
+    tail_io_ctx.locator_set_key(read_obj.loc);
   }
 
   ldout(cct, 20) << "rados->read obj-ofs=" << ofs << " read_ofs=" << read_ofs << " read_len=" << read_len << dendl;
   op.read(read_ofs, read_len, pbl, NULL);
 
-  r = state.io_ctx.operate(read_obj.oid, &op, NULL);
+  if (reading_from_head)
```
Comment (yehudasa) [inline src/rgw/rgw_rados.cc]: @fangyuxiangGL why do we need a special handling here? The code already knows how to deal with different head and tail pools. Everyting should be in the manifest. Maybe you missed updating the manifest correctly?

File: src/rgw/rgw_common.h (hunk 1 above)
Comment (fangyuxiangGL) [inline src/rgw/rgw_common.h]: you are right, already decoupled from rgw_obj, and move it to rgw_obj_select.

File: src/rgw/rgw_rados.cc (hunk 2 above)
Comment (fangyuxiangGL) [inline src/rgw/rgw_rados.cc]: reverted the modification, now rgw_get_obj_data_pool is not rewritten.

File: src/rgw/rgw_rados.cc (hunk 3 above)
Comment (fangyuxiangGL) [inline src/rgw/rgw_rados.cc]: it is not related to manifest.

we have two pools now, and will also have two rados pool ioctx. so, which rados ioctx should we use when code reach here? we can do the choice according to whether it is head object or not.

**Comment by fangyuxiangGL** [comment]
jenkins test this please

**Comment by fangyuxiangGL** [comment]
jenkins test this please

**Comment by fangyuxiangGL** [comment]
jenkins test this please

File: src/rgw/rgw_common.h (hunk 1 above)
Comment (mikulely) [inline src/rgw/rgw_common.h]: @fangyuxiangGL Maybe not necessary to introduce this, we can use the offset to determine whether we are in the head/tail.

File: src/rgw/rgw_rados.cc (hunk 3 above)
Comment (mikulely) [inline src/rgw/rgw_rados.cc]: @fangyuxiangGL nice catch!

File: src/rgw/rgw_common.h (hunk 1 above)
Comment (mattbenjamin) [inline src/rgw/rgw_common.h]: @mikulely I actually prefer the explicitness of head_obj, and don't mind the minimal cost of computing it early.

Comment (fangyuxiangGL) [inline src/rgw/rgw_common.h]: @mattbenjamin @mikulely 

please see the update codes, I had changed the schem according @yehudasa 's comments.

File: src/rgw/rgw_op.cc [hunk 4]
```diff
@@ -2992,6 +2992,8 @@ int RGWPutObjProcessor_Multipart::do_complete(size_t accounted_size,
   head_obj_op.meta.delete_at = delete_at;
   head_obj_op.meta.zones_trace = zones_trace;
 
+  op_target.set_in_tail(true);
```
Comment (Liuchang0812) [inline src/rgw/rgw_op.cc]: why not atomic upload use tail data pool ?

File: src/rgw/rgw_rados.h [hunk 5]
```diff
@@ -1323,6 +1347,23 @@ struct RGWZoneParams : RGWSystemMetaObj {
     }
     return true;
   }
+
+  bool get_tail_data_pool(const string& placement_id, const rgw_obj& obj, rgw_pool *pool) const {
+    const rgw_data_placement_target& explicit_placement = obj.bucket.explicit_placement;
+    if (!explicit_placement.data_pool.empty()) {
```
Comment (Liuchang0812) [inline src/rgw/rgw_rados.h]: it should be 
```
if (!explicit_placement.data_tail_pool.empty()) {
```

File: src/rgw/rgw_rados.cc [hunk 6]
```diff
@@ -3385,7 +3404,26 @@ int RGWRados::get_max_chunk_size(const string& placement_rule, const rgw_obj& ob
...
+  
+  int r = get_max_chunk_size(pool, &chunk_size);
+  if (r < 0)
+    return r;
+
+  r = get_max_chunk_size(tail_pool, &tail_chunk_size);
+  if (r < 0)
+    return r;
+
+  *max_chunk_size = MIN(chunk_size, tail_chunk_size);
```
Comment (Liuchang0812) [inline src/rgw/rgw_rados.cc]: max_chunk_size is wrong when there are 2 EC data pool.

Comment (Liuchang0812) [reply src/rgw/rgw_rados.cc]: > max_chunk_size is wrong when there are 2 EC data pool. `max_chunk_size` should be aligned with those two ec pools, we have to return least common multiple here.

sorry. `data pool` must be rep pool, not EC pool.

**Pull request closed without merging**